### PR TITLE
Return error if there are multiple same name params

### LINKED ISSUE #2016: Pipeline state stuck in running even when task is finished
# Expected Behavior

The pipeline should move to the next task

# Actual Behavior

The pipeline does not initiate the next step

# Steps to Reproduce the Problem

The problem started when I added param to the task

# Additional Info

Here is the pipelinerun, as you can see it state that the task build-and-push image is still running: 

```
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"tekton.dev/v1alpha1","kind":"Pipeline","metadata":{"annotations":{},"name":"build-push-and-deploy","namespace":"blu"},"spec":{"params":[{"default":"registry.gitlab.com/blubybs2/blu-api","description":"image url without tag","name":"image","type":"string"},{"default":"blu-api","description":"Name of the deployment resource, ex blu-api-stg","name":"deploymentName","type":"string"},{"default":"blu","description":"namespace to apply the path","name":"namespace","type":"string"},{"default":"blu-api","name":"project","type":"string"},{"name":"commitHash","type":"string"}],"resources":[{"name":"workspace","type":"git"}],"tasks":[{"name":"build","resources":{"inputs":[{"name":"workspace","resource":"workspace"}],"outputs":[{"name":"workspace","resource":"workspace"}]},"taskRef":{"name":"build-gradle"}},{"conditions":[{"conditionRef":"is-master","resources":[{"name":"workspace","resource":"workspace"}]}],"name":"build-and-push-image","params":[{"name":"image","value":"$(params.image)"}],"resources":{"inputs":[{"from":["build"],"name":"workspace","resource":"workspace"}]},"taskRef":{"name":"build-and-push-image"}},{"name":"deploy","params":[{"name":"image","value":"${params.image}"},{"name":"deploymentName","value":"${params.deploymentName}"},{"name":"namespace","value":"${params.namespace}"},{"name":"image","value":"${params.image}"}],"resources":{"inputs":[{"from":["build"],"name":"workspace","resource":"workspace"}]},"runAfter":["build-and-push-image"],"taskRef":{"name":"deploy-production"}}]}}
  creationTimestamp: "2020-02-06T22:10:30Z"
  generateName: blu-d9dqt-r-
  generation: 1
  labels:
    reruns: ""
    tekton.dev/eventlistener: blu-api-listener
    tekton.dev/pipeline: build-push-and-deploy
    tekton.dev/trigger: ""
    tekton.dev/triggers-eventid: fgtlz
  name: blu-d9dqt-r-thqn2
  namespace: blu
  resourceVersion: "24370027"
  selfLink: /apis/tekton.dev/v1alpha1/namespaces/blu/pipelineruns/blu-d9dqt-r-thqn2
  uid: 637ca2b5-9721-496e-9bd8-b205e2a34c53
spec:
  params:
  - name: project
    value: "10593334"
  - name: namespace
    value: blu
  - name: commitHash
    value: fce2d06c7b27b13ec9e6c98cbfb74fa782256cbf
  - name: image
    value: registry.gitlab.com/blubybs2/blu-paymee:fce2d06c7b27b13ec9e6c98cbfb74fa782256cbf
  pipelineRef:
    name: build-push-and-deploy
  podTemplate: {}
  resources:
  - name: workspace
    resourceRef: {}
    resourceSpec:
      params:
      - name: revision
        value: refs/heads/develop
      - name: url
        value: git@gitlab.com:blubybs2/blu-paymee.git
      type: git
  serviceAccountName: build-bot
  timeout: 1h0m0s
status:
  conditions:
  - lastTransitionTime: "2020-02-06T22:10:30Z"
    message: Not all Tasks in the Pipeline have finished executing
    reason: Running
    status: Unknown
    type: Succeeded
  startTime: "2020-02-06T22:10:30Z"
  taskRuns:
    blu-d9dqt-r-thqn2-build-and-push-image-xr5zg:
      conditionChecks:
        blu-d9dqt-r-thqn2-build-and-push-image-xr5zg-is-master-rpmth:
          conditionName: is-master
          status:
            check:
              terminated:
                containerID: docker://c8a6afc11fb94836632f12db1338f688afe8fb9ac3034d0c77adc12c758fcc38
                exitCode: 0
                finishedAt: "2020-02-06T22:13:30Z"
                reason: Completed
                startedAt: "2020-02-06T22:13:29Z"
            completionTime: "2020-02-06T22:13:30Z"
            conditions:
            - lastTransitionTime: "2020-02-06T22:13:30Z"
              message: All Steps have completed executing
              reason: Succeeded
              status: "True"
              type: Succeeded
            podName: blu-d9dqt-r-thqn2-build-and-push-image-xr5zg-is-master-rp-pmc4v
            startTime: "2020-02-06T22:13:19Z"
      pipelineTaskName: build-and-push-image
      status:
        conditions:
        - lastTransitionTime: "2020-02-06T22:13:55Z"
          message: Not all Steps in the Task have finished executing
          reason: Running
          status: Unknown
          type: Succeeded
        podName: blu-d9dqt-r-thqn2-build-and-push-image-xr5zg-pod-zcslv
        startTime: "2020-02-06T22:13:30Z"
        steps:
        - container: step-build-and-push-image
          imageID: docker-pullable://gcr.io/kaniko-project/executor@sha256:bcfcad6b54e9368079232094ed9f8332e5bc642aed77a6b0c1f15fee2b8b13df
          name: build-and-push-image
          running:
            startedAt: "2020-02-06T22:13:54Z"
        - container: step-create-dir-workspace-pzhvs
          imageID: docker-pullable://busybox@sha256:edafc0a0fb057813850d1ba44014914ca02d671ae247107ca70c94db686e7de6
          name: create-dir-workspace-pzhvs
          terminated:
            containerID: docker://c671fc9478dd08382d6a23541f6db2f3b503c40f07ac9d9b0dc351e9d677ee38
            exitCode: 0
            finishedAt: "2020-02-06T22:13:56Z"
            reason: Completed
            startedAt: "2020-02-06T22:13:56Z"
        - container: step-source-copy-workspace-mb6r8
          imageID: docker-pullable://busybox@sha256:edafc0a0fb057813850d1ba44014914ca02d671ae247107ca70c94db686e7de6
          name: source-copy-workspace-mb6r8
          terminated:
            containerID: docker://0eb838bb0fbb1619497cf0fe66f99dfe225d7a422a4b708b267f47f495de3a23
            exitCode: 0
            finishedAt: "2020-02-06T22:13:57Z"
            reason: Completed
            startedAt: "2020-02-06T22:13:57Z"
    blu-d9dqt-r-thqn2-build-hcccm:
      pipelineTaskName: build
      status:
        completionTime: "2020-02-06T22:13:19Z"
        conditions:
        - lastTransitionTime: "2020-02-06T22:13:19Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        podName: blu-d9dqt-r-thqn2-build-hcccm-pod-dq9nf
        resourcesResult:
        - digest: ""
          key: commit
          name: ""
          resourceRef: {}
          value: 9ae35fbe401a0005596d8d00c8fade1340a2ed7d
        startTime: "2020-02-06T22:10:30Z"
        steps:
        - container: step-build-gradle
          imageID: docker-pullable://openjdk@sha256:4fb568c15fa164d4ccf64a5db54ab73d7d395b06ca8b8af07e3c0793cfa07e55
          name: build-gradle
          terminated:
            containerID: docker://b948472336d2c210505f69be3e35292c5172524c83ead14b6ac9d5afa518f64a
            exitCode: 0
            finishedAt: "2020-02-06T22:13:14Z"
            reason: Completed
            startedAt: "2020-02-06T22:11:04Z"
        - container: step-upload-coverage
          imageID: docker-pullable://mrnonz/alpine-git-curl@sha256:4dda2a964fa309aaf291ee4b633a6cb790b20599d008f6294e9e08edf773a5c8
          name: upload-coverage
          terminated:
            containerID: docker://ce6dcff2c6c8314a1465c0ef5432b1fb928c03aff1a20eb87344edd9096d0ef5
            exitCode: 0
            finishedAt: "2020-02-06T22:13:17Z"
            reason: Completed
            startedAt: "2020-02-06T22:13:14Z"
        - container: step-git-source-workspace-dd2cd
          imageID: docker-pullable://gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init@sha256:5b30b2e31583dde58285cfa6e024f4a33582c132c3126b7e8880fd4c51882dd1
          name: git-source-workspace-dd2cd
          terminated:
            containerID: docker://085eaf154add7ae1b6c3921d7fad7a451b32736a0baa40127c768242702332ce
            exitCode: 0
            finishedAt: "2020-02-06T22:11:04Z"
            message: '[{"name":"","digest":"","key":"commit","value":"9ae35fbe401a0005596d8d00c8fade1340a2ed7d","resourceRef":{}}]'
            reason: Completed
            startedAt: "2020-02-06T22:11:01Z"
        - container: step-source-copy-workspace-ml4kp
          imageID: docker-pullable://busybox@sha256:edafc0a0fb057813850d1ba44014914ca02d671ae247107ca70c94db686e7de6
          name: source-copy-workspace-ml4kp
          terminated:
            containerID: docker://e20187883fdf8619d85db288cc9fca45e444217c1b0c5759c5703cfcc2f0af7a
            exitCode: 0
            finishedAt: "2020-02-06T22:13:18Z"
            reason: Completed
            startedAt: "2020-02-06T22:13:18Z"
        - container: step-source-mkdir-workspace-bk4nk
          imageID: docker-pullable://busybox@sha256:edafc0a0fb057813850d1ba44014914ca02d671ae247107ca70c94db686e7de6
          name: source-mkdir-workspace-bk4nk
          terminated:
            containerID: docker://a6e2c1a665351ecce6713f571c408e0d6fbb50d629b4ba31171f175820e4923f
            exitCode: 0
            finishedAt: "2020-02-06T22:13:17Z"
            reason: Completed
            startedAt: "2020-02-06T22:13:17Z"
        - container: step-create-dir-workspace-qbfnx
          imageID: docker-pullable://busybox@sha256:edafc0a0fb057813850d1ba44014914ca02d671ae247107ca70c94db686e7de6
          name: create-dir-workspace-qbfnx
          terminated:
            containerID: docker://0070494d8d4be6b9c77ec7878786cb8ff5b7dca25e8c2cc8f0f44018b353ec5e
            exitCode: 0
            finishedAt: "2020-02-06T22:11:00Z"
            reason: Completed
            startedAt: "2020-02-06T22:11:00Z"

```

And here is the taskRun, as you can see the task has finished with success:

```
apiVersion: tekton.dev/v1alpha1
kind: TaskRun
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"tekton.dev/v1alpha1","kind":"Task","metadata":{"annotations":{},"name":"build-and-push-image","namespace":"blu"},"spec":{"inputs":{"params":[{"description":"Image string url without tag, ex registry.gitlab.com/blubybs2/blu-api","name":"image","type":"string"}],"resources":[{"name":"workspace","targetPath":"./","type":"git"}]},"steps":[{"env":[{"name":"DOCKER_CONFIG","value":"/builder/home/.docker/"}],"image":"gcr.io/kaniko-project/executor:debug","name":"build-and-push-image","script":"#!/busybox/sh\nset -e\nset -x\n\n/kaniko/executor \\\n --dockerfile=/workspace/Dockerfile \\\n --destination=$(inputs.params.image) \\\n --context=/workspace \\\n --verbosity=debug\n"}]}}
  creationTimestamp: "2020-02-06T22:13:30Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: tekton-pipelines
    reruns: ""
    tekton.dev/eventlistener: blu-api-listener
    tekton.dev/pipeline: build-push-and-deploy
    tekton.dev/pipelineRun: blu-d9dqt-r-thqn2
    tekton.dev/pipelineTask: build-and-push-image
    tekton.dev/task: build-and-push-image
    tekton.dev/trigger: ""
    tekton.dev/triggers-eventid: fgtlz
  name: blu-d9dqt-r-thqn2-build-and-push-image-xr5zg
  namespace: blu
  ownerReferences:
  - apiVersion: tekton.dev/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: PipelineRun
    name: blu-d9dqt-r-thqn2
    uid: 637ca2b5-9721-496e-9bd8-b205e2a34c53
  resourceVersion: "24370161"
  selfLink: /apis/tekton.dev/v1alpha1/namespaces/blu/taskruns/blu-d9dqt-r-thqn2-build-and-push-image-xr5zg
  uid: 40607299-839a-48e9-8678-1c65d0cd3b42
spec:
  inputs:
    params:
    - name: image
      value: registry.gitlab.com/blubybs2/blu-paymee:fce2d06c7b27b13ec9e6c98cbfb74fa782256cbf
    resources:
    - name: workspace
      paths:
      - /pvc/build/workspace
      resourceSpec:
        params:
        - name: revision
          value: refs/heads/develop
        - name: url
          value: git@gitlab.com:blubybs2/blu-paymee.git
        type: git
  outputs: {}
  podTemplate: {}
  serviceAccountName: build-bot
  taskRef:
    kind: Task
    name: build-and-push-image
  timeout: 1h0m0s
status:
  completionTime: "2020-02-06T22:14:20Z"
  conditions:
  - lastTransitionTime: "2020-02-06T22:14:20Z"
    message: All Steps have completed executing
    reason: Succeeded
    status: "True"
    type: Succeeded
  podName: blu-d9dqt-r-thqn2-build-and-push-image-xr5zg-pod-zcslv
  startTime: "2020-02-06T22:13:30Z"
  steps:
  - container: step-build-and-push-image
    imageID: docker-pullable://gcr.io/kaniko-project/executor@sha256:bcfcad6b54e9368079232094ed9f8332e5bc642aed77a6b0c1f15fee2b8b13df
    name: build-and-push-image
    terminated:
      containerID: docker://64490186cc54abd4a47ee7a768f5f8ae85349287d3ff4420dd80a975773bccf0
      exitCode: 0
      finishedAt: "2020-02-06T22:14:19Z"
      reason: Completed
      startedAt: "2020-02-06T22:13:58Z"
  - container: step-create-dir-workspace-pzhvs
    imageID: docker-pullable://busybox@sha256:edafc0a0fb057813850d1ba44014914ca02d671ae247107ca70c94db686e7de6
    name: create-dir-workspace-pzhvs
    terminated:
      containerID: docker://c671fc9478dd08382d6a23541f6db2f3b503c40f07ac9d9b0dc351e9d677ee38
      exitCode: 0
      finishedAt: "2020-02-06T22:13:56Z"
      reason: Completed
      startedAt: "2020-02-06T22:13:56Z"
  - container: step-source-copy-workspace-mb6r8
    imageID: docker-pullable://busybox@sha256:edafc0a0fb057813850d1ba44014914ca02d671ae247107ca70c94db686e7de6
    name: source-copy-workspace-mb6r8
    terminated:
      containerID: docker://0eb838bb0fbb1619497cf0fe66f99dfe225d7a422a4b708b267f47f495de3a23
      exitCode: 0
      finishedAt: "2020-02-06T22:13:57Z"
      reason: Completed
      startedAt: "2020-02-06T22:13:57Z"

```

### PR DESCRIPTION
**Pull request by NavidZ**
# Changes
Right now pipeline accepts multiple parameters with
the same name. This sometime causes the pipeline to
not finish.

This change adds a simple check and returns error
when there are multiple parameters with the same
name in the spec.params.

Fixes #2016

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
After this change it is not allowed to have multiple parameters with the same name
within pipeline spec.
```

### CODE REVIEW

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go [hunk 1]
```diff
@@ -916,6 +916,14 @@ func TestValidatePipelineParameterVariables_Failure(t *testing.T) {
...
 			}},
 		}},
+	}, {
+		name: "multiple parameters with the same name",
+		params: []ParamSpec{{
+			Name: "baz", Type: ParamTypeString, Default: &ArrayOrString{Type: ParamTypeString, StringVal: "astring"},
+		}, {
+			Name: "baz", Type: ParamTypeString, Default: &ArrayOrString{Type: ParamTypeString, StringVal: "astring"},
+		}},
+		tasks: []PipelineTask{},
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:926]: can we add a single valid task such that this test confirms that the pipeline validation does not fail because of no tasks but fails because of duplicate params?

Comment (NavidZ) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:926]: The empty task list is valid though at least in the current state. So there should be no difference between an empty list and one valid task. Considering that, if you still prefer to have a single valid task here let me know and I'll add it.

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go [hunk 2]
```diff
@@ -341,6 +341,9 @@ func validatePipelineParameterVariables(tasks []PipelineTask, params []ParamSpec
 			}
 		}
 
+		if _, ok := parameterNames[p.Name]; ok {
```
Comment (pritidesai) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:344]: is it possible to add the same condition for `arrayParameterNames`?

Also, we are still allowing params of different types having same `p.Name`, since we are updating release notes, can we add the check and not permit duplicate param names having different types?

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go (hunk 1 above)
Comment (pritidesai) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:926]: yes please, so that it isolates the tests being done 🙏

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go (hunk 2 above)
Comment (NavidZ) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:344]: This logic catches that issue too as far as I can see since this line does execute for all params regardless of their type. I added more tests to show the effect of it and the fact that it works in that case as well. Let me know if I understood what you said correctly or if you wanted me to do something else.

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go (hunk 1 above)
Comment (NavidZ) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go]: done

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go (hunk 2 above)
Comment (pritidesai) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:344]: yeah I see, `parameterNames` holds all parameters whereas `arrayParameterNames` is a subset of `parameterNames` and only holds array params. Sorry for confusion 🙏

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go (hunk 1 above)
Comment (pritidesai) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go]: thanks for adding more tests, it now has some coverage for the error being added.

Before:

Screen Shot 2020-06-16 at 9 39 56 AM

Now:

Screen Shot 2020-06-16 at 9 40 22 AM
